README.md
# SimpleBench

SimpleBench is a quick benchmarking tool I wrote to evaluate the performance of the systems I came across. It tests only the most basic arithmetic operations for integers and floating points besides memory bandwitdth.

![Image](screenshots/1.png)

It supports the following options:
```
--show-gui [on/off]: enable/disable the text feedback during the tests
--st-test [on/off]: enable/disable the singlethread test
--mt-test [on/off]: enable/disable the multithread test
--threads [integer greater than zero]: sets the number of threads to run with
--hardware-level [integer greater than zero]: chooses how taxing the benchmark will be
      1: 32MB of RAM, 64KB of cache, lenght 1
      2: 64MB of RAM, 128KB of cache, lenght 4
      3: 128MB of RAM, 256KB of cache, lenght 16
      4: 512MB of RAM, 1MB of cache, lenght 64
      5: 2GB of RAM, 4MB of cache, lenght 256 (this is the current default)
      6: 8GB of RAM, 16MB of cache, lenght 1024 (HEDT users may want to use this one)
       :
      * every increase of 1 in the hardware level quadruples the requirements
      * there is some overhead in RAM requirement, you should have at least the amount required plus 5% FREE
--version: shows the benchmark version (not the program version)
--build: shows the program version, changes here have no effect on the score system
--help: shows this screen
```

SimpleBench is not a stress test and not is not a reliable benchmarking tool for any kind of advanced features the processor may have. 

Also, the algorithm is not complex so the compiler optimizations will basically skip all work it should have done. If you want to build from the sources, disable any optimizations to have a meaningful score. There are scripts to build with the correct compiler flags included in the source folder too.

These are the systems I benchmarked so far (I plan to include more systems in the future).

<table id="Scores">
    <tr> <!--Table header-->
        <th>ST Score</th>
        <th>MT Score</th>
        <th>Multiplier</th>
        <th>CPU Vendor</th>
        <th>CPU Model</th>
        <th>CPU Cores</th>
        <th>CPU Threads</th>
        <th>CPU All Core Turbo</th>
        <th>CPU Single Core Turbo</th>
        <th>CPU Cache</th>
        <th>RAM Vendor</th>
        <th>RAM Model</th>
        <th>RAM Amount</th>
        <th>RAM Channels</th>
        <th>RAM Frequency</th>
        <th>RAM Timings</th>
        <th>Operating System</th>
        <th>OS Update</th>
        <th>OS Kernel</th>
        <th>Hardware Level</th>
        <th>Version</th>
        <th>Build</th>
        <th>CLI Options</th>
    </tr>
    <tr> <!--Intel Xeon E3 1280-->
        <td>100.14</td><!--ST Score-->
        <td>426.92</td><!--MT Score-->
        <td>4.26</td><!--Multiplier-->
        <td>Intel</td><!--CPU Vendor-->
        <td>Xeon E3 1280</td><!--CPU Model-->
        <td>4</td><!--CPU Cores-->
        <td>8</td><!--CPU Threads-->
        <td>3600MHz</td><!--CPU All Core Turbo-->
        <td>3900MHz</td><!--CPU Single Core Turbo-->
        <td>8MB</td><!--CPU Cache-->
        <td>HyperX</td><!--RAM Vendor-->
        <td>Fury</td><!--RAM Model-->
        <td>8GB</td><!--RAM Amount-->
        <td>1</td><!--RAM Channels-->
        <td>1333MHz</td><!--RAM Frequency-->
        <td>8-8-8-18</td><!--RAM Timings-->
        <td>Arch Linux</td><!--Operating System-->
        <td>2019-03</td><!--OS Update-->
        <td>Linux 5.0</td><!--OS Kernel-->
        <td>5</td><!--Hardware Level-->
        <td>A0</td><!--Version-->
        <td>1-Linux-64 Absolution</td><!--Build-->
        <td></td><!--CLI Options-->
    </tr>
    <tr> <!--Intel Celeron 430-->
        <td>30.73</td><!--ST Score-->
        <td>30.73</td><!--MT Score-->
        <td>1.00</td><!--Multiplier-->
        <td>Intel</td><!--CPU Vendor-->
        <td>Celeron 430</td><!--CPU Model-->
        <td>1</td><!--CPU Cores-->
        <td>1</td><!--CPU Threads-->
        <td>1800MHz</td><!--CPU All Core Turbo-->
        <td>1800MHz</td><!--CPU Single Core Turbo-->
        <td></td><!--CPU Cache-->
        <td>Markvision</td><!--RAM Vendor-->
        <td></td><!--RAM Model-->
        <td>2GB</td><!--RAM Amount-->
        <td>1</td><!--RAM Channels-->
        <td>667MHz</td><!--RAM Frequency-->
        <td></td><!--RAM Timings-->
        <td>Debian 9</td><!--Operating System-->
        <td>2019-03</td><!--OS Update-->
        <td>Linux 4.9</td><!--OS Kernel-->
        <td>4</td><!--Hardware Level-->
        <td>A0</td><!--Version-->
        <td>1-Linux-64 Absolution</td><!--Build-->
        <td>--hardware-level 4</td><!--CLI Options-->
    </tr>
    <tr> <!--Template-->
        <td></td><!--ST Score-->
        <td></td><!--MT Score-->
        <td></td><!--Multiplier-->
        <td></td><!--CPU Vendor-->
        <td></td><!--CPU Model-->
        <td></td><!--CPU Cores-->
        <td></td><!--CPU Threads-->
        <td></td><!--CPU All Core Turbo-->
        <td></td><!--CPU Single Core Turbo-->
        <td></td><!--CPU Cache-->
        <td></td><!--RAM Vendor-->
        <td></td><!--RAM Model-->
        <td></td><!--RAM Amount-->
        <td></td><!--RAM Channels-->
        <td></td><!--RAM Frequency-->
        <td></td><!--RAM Timings-->
        <td></td><!--Operating System-->
        <td></td><!--OS Update-->
        <td></td><!--OS Kernel-->
        <td></td><!--Hardware Level-->
        <td></td><!--Version-->
        <td></td><!--Build-->
        <td></td><!--CLI Options-->
    </tr>
</table>

## [1deterministic](https://github.com/1deterministic), 2019


changelog.txt
# = Program build | $ = Benchmark version

# Build 1
Initial release

$ Benchmark A0
Initial release

main.c
#include "simplebench.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>

#ifdef __linux__
    #include <unistd.h>
    
#elif __MINGW64__ || __MINGW32__ || _WIN32
#endif

// normalizes the score
float handicap;
// size of the matrices being used
int alu_matrix_size;
int fpu_matrix_size;
int mem_matrix_size;
// size of the tasks
int alu_job_size;
int fpu_job_size;
int mem_job_size;

int main(int argc, char** argv) {
    MsgCode code = SUCCESS;

    float singlethread_score = 0.0;
    float multithread_score = 0.0;

    // default configuration
    #ifdef __linux__
        int threads = sysconf(_SC_NPROCESSORS_ONLN);

    #elif __MINGW64__ || __MINGW32__ || _WIN32
        SYSTEM_INFO info;
        GetSystemInfo(&info);
        int threads = info.dwNumberOfProcessors;

    #endif
    
    int hardware_level = DEFAULT_CONFIG_HARDWARE;
    bool show_gui = DEFAULT_SHOW_GUI;
    bool st_test = DEFAULT_ST_TEST;
    bool mt_test = DEFAULT_MT_TEST;

    code = get_cli_options(argc, argv, &show_gui, &st_test, &mt_test, &hardware_level, &threads);
    if (code) {printf("%s\n", get_string(code)); return code;}

    load_test_config(hardware_level);
    
    // gets the scores fot the single and multithreaded tests
    if (st_test) {
        code = test_system(&singlethread_score, 1, handicap, show_gui);
        if (code) {printf("%s\n", get_string(code)); return code;}
    }
    
    if (mt_test) {
        // skip the multithread test if the system only has 1
        if (threads > 1) {
            code = test_system(&multithread_score, threads, handicap, show_gui);
            if (code) {printf("%s\n", get_string(code)); return code;}
        } else {
            multithread_score = singlethread_score;
        }
    }

    show_score(singlethread_score, multithread_score, threads, hardware_level);
    return code;
}

// reads the options received via CLI
MsgCode get_cli_options(int argc, char** argv, bool* show_gui, bool* st_test, bool* mt_test, int* hardware_level, int* threads) {
    for (int i = 1; i < argc; i++) {
        // --show-gui
        if (strcmp(argv[i], CLI_SHOW_GUI) == 0) {
            // check if the next argument exists before trying to access
            if (i + 1 < argc) {
                i++;
                // if the option is "on"
                if (strcmp(argv[i], CLI_ON) == 0) {
                    *show_gui = true;
                }
                // if the option is "off"
                else if (strcmp(argv[i], CLI_OFF) == 0) {
                    *show_gui = false;
                }
                // next argument is not "on"/"off"
                else {
                    return MSG_GET_CLI_OPTIONS_INVALID_ONOFF;
                }
                // no need to test the other if's, goes to the next argument
                continue;
            } else {
                // did not find "on"/"off", reports the error
                return MSG_GET_CLI_OPTIONS_MISSING_ONOFF;
            }
        }

        // --st-test
        if (strcmp(argv[i], CLI_ST_TEST) == 0) {
            // check if the next argument exists before trying to access
            if (i + 1 < argc) {
                i++;
                // if the option is "on"
                if (strcmp(argv[i], CLI_ON) == 0) {
                    *st_test = true;
                }
                // if the option is "off"
                else if (strcmp(argv[i], CLI_OFF) == 0) {
                    *st_test = false;
                }
                // next argument is not "on"/"off"
                else {
                    return MSG_GET_CLI_OPTIONS_INVALID_ONOFF;
                }
                // no need to test the other if's, goes to the next argument
                continue;
            } else {
                // did not find "on"/"off", reports the error
                return MSG_GET_CLI_OPTIONS_MISSING_ONOFF;
            }
        }

        // --mt-test
        if (strcmp(argv[i], CLI_MT_TEST) == 0) {
            // check if the next argument exists before trying to access
            if (i + 1 < argc) {
                i++;
                // if the option is "on"
                if (strcmp(argv[i], CLI_ON) == 0) {
                    *mt_test = true;
                }
                // if the option is "off"
                else if (strcmp(argv[i], CLI_OFF) == 0) {
                    *mt_test = false;
                }
                // next argument is not "on"/"off"
                else {
                    return MSG_GET_CLI_OPTIONS_INVALID_ONOFF;
                }
                // no need to test the other if's, goes to the next argument
                continue;
            } else {
                // did not find "on"/"off", reports the error
                return MSG_GET_CLI_OPTIONS_MISSING_ONOFF;
            }
        }

        // --threads
        else if (strcmp(argv[i], CLI_THREADS) == 0) {
            // check if the next argument exists before trying to access
            if (i + 1 < argc) {
                i++;
                // converts the text to an integer value
                *threads = atoi(argv[i]);
                // only values greater than zero are allowed
                // also, atoi will return 0 if the text is not an integer
                if (*threads <= 0) {
                    return MSG_GET_CLI_OPTIONS_INVALID_INT;
                }   
                // no need to test the other if's, goes to the next argument 
                continue;
            } else {
                // did not find a valid value, reports the error
                return MSG_GET_CLI_OPTIONS_MISSING_INT;
            }
        }

        // --hardware-level
        if (strcmp(argv[i], CLI_HARDWARE_LEVEL) == 0) {
            // check if the next argument exists before trying to access
            if (i + 1 < argc) {
                i++;
                // converts the text to an integer value
                *hardware_level = atoi(argv[i]);
                // only values greater than zero are allowed
                // also, atoi will return 0 if the text is not an integer
                if (*hardware_level <= 0) {
                    return MSG_GET_CLI_OPTIONS_INVALID_INT;
                }
                // no need to test the other if's, goes to the next argument 
                continue;
            } else {
                // did not find a valid value, reports the error
                return MSG_GET_CLI_OPTIONS_MISSING_INT;
            }
        }

        // --version
        else if (strcmp(argv[i], CLI_VERSION) == 0) {
            return MSG_GET_CLI_OPTIONS_VERSION;
        }

        // --build
        else if (strcmp(argv[i], CLI_BUILD) == 0) {
            return MSG_GET_CLI_OPTIONS_BUILD;
        }

        // --help
        else if (strcmp(argv[i], CLI_HELP) == 0) {            
            return MSG_HELP_TEXT;
        }

        // not recognized option
        else {
            return MSG_GET_CLI_OPTIONS_UNKNOWN_OPTION;
        }
    }

    return SUCCESS;
}

// shows the system score
void show_score(float singlethread_score, float multithread_score, int threads, int hardware_level) {
    printf("%-20s: %20s\n", get_string(MSG_MAIN_SHOW_SCORE_SIMPLEBENCH_HEADER), BUILD_CODENAME);
    printf("%-20s: %20s\n", get_string(MSG_MAIN_SHOW_SCORE_SIMPLEBENCH_VERSION), BENCHMARK_VERSION);
    printf("%-20s: %20s\n", get_string(MSG_MAIN_SHOW_SCORE_SIMPLEBENCH_BUILD), BUILD_VERSION);
    printf("%-20s: %20d\n", get_string(MSG_MAIN_SHOW_SCORE_HARDWARE_LEVEL), hardware_level);
    printf("%-20s: %20.2f\n", get_string(MSG_MAIN_SHOW_SCORE_SINGLETHREAD_SCORE), singlethread_score);
    printf("%-20s: %20.2f\n", get_string(MSG_MAIN_SHOW_SCORE_MULTITHREAD_SCORE), multithread_score);
    printf("%-20s: %20.2f\n", get_string(MSG_MAIN_SHOW_SCORE_MULTIPLIER), (singlethread_score > 0.0) ? multithread_score / singlethread_score : 0.0);
}

// loads the test config
void load_test_config(int config) {
    // the sizes of the matrices changes with the hardware level used
    // every step up on hardware level doubles the sizes of the matrices
    // this increase the amount of work by 4 times, so the handicap is then increased by 4
    handicap = 0.00390625 * pow(4, config - 1);
    alu_matrix_size = 32 * pow(2, config - 1);
    fpu_matrix_size = 32 * pow(2, config - 1);
    mem_matrix_size = 1024 * pow(2, config - 1);
    alu_job_size = 8192;
    fpu_job_size = 256; // this limits thread count, improve on future
    mem_job_size = 131072;
}